src/aiida_quantumespresso/data/hubbard_structure.py

```diff
@@ -0,0 +1,996 @@
+        """
+        This finds the index of the atom within the unitcell from its image position.
+        
+        :param position: (3,) array list describing the position of the atom in Cartesian coordinates.
+        """
+        ase = self.get_ase()
+        
+        cell = ase.cell
+        inv_cell = np.linalg.inv(cell)
+        positions = ase.positions
+        
+        for index, uc_position in enumerate(positions):
+            diff = (uc_position-np.array(position))
+            translation = -np.dot(diff, inv_cell)
+            translation_int = np.rint(translation)
+            if np.all(np.isclose(translation, translation_int, thr)):
+                break
+        
+        return index, translation_int
+    
+    def get_images_distance_and_translation(self, atom_index_i: int, atom_index_j: int):
+        """Returns the distance and translations between atom i and the 3x3x3 supercell
+        images of atom j in ascending order in distance.
+        
+        :param atom_index_*: number indicating the atom * in the unitcell; it ranges between 0 and (number of atoms -1)
+        """
+        translations = self._get_standard_translations()
+        distances = []
+        
+        for translation in translations:
+            distance_ij = self.atomic_distance(atom_index_i, atom_index_j, [0,0,0], translation)
+            distances.append(distance_ij)
+        
+        sorting = np.argsort(distances)
+        translations = [translations[i] for i in sorting]
+        distances.sort()
+        
+        return distances, translations
+        
+    @property
+    def hubbard_projectors(self):
+        """Get the Hubbard projectors.
+        
+        :return: string
+        """
+        return self.get_attribute('hubbard_projectors')
+    
+    @hubbard_projectors.setter
+    def hubbard_projectors(self, value: str):
+        """Set the Hubbard projectors."""
+        if isinstance(value, str):
+            self.set_attribute('hubbard_projectors', deepcopy(value))
+        else:
+            raise ValueError(f"type <{type(value)}> is not correct; only class <{type(str)}> is accepted")
+        
+    @property
+    def hubbard_parameters(self):
+        """Get the Hubbard parameters.
+        
+        :return: list of list with 7 values describing the Hubbard parameters.
+        """
+        try:
+            the_hps = self.get_attribute('hubbard_parameters')
+        except AttributeError:
+            the_hps = []
+        return the_hps
+    
+    @hubbard_parameters.setter
+    def hubbard_parameters(self, hp_parameters: list):
+        """Set the full Hubbard parameters from a list (overrides completely the current ones)."""
+        self._set_hubbard_parameters(hp_parameters)
+    
+    def _set_hubbard_parameters(self, hubbard_parameters: list):
+        """Check and set the full list of Hubbard parameters."""
+        num_atoms = len(self.sites)
+        if isinstance(hubbard_parameters, list):
```

Comment:
Just saw this in passing by and thought I'd give a tip. Instead of doing this paradigm:
```python
if some_statement:
    # some code
else:
    raise Exception
```
do the following
```python
if not statement:
    raise Exception

# some code
```
The code will be much more readable as you won't have so many levels of nesting